rel/m_lib.c

```diff
@@ -1,371 +1,880 @@
+ *
+ * @param pValue Pointer to the value to be updated.
+ * @param target Target value to chase.
+ * @param step Step value to be added or subtracted from the current value.
+ * @return TRUE if the value reaches the target, FALSE otherwise.
+ */
+extern int chase_f(f32* const pValue, const f32 target, f32 step) {
+  if (step) {
+    if (*pValue > target) {
+      step = -step;
+    }
 
     *pValue += step;
-    if (((s16)(*pValue - limit) * (s16)(prev - limit)) <= 0) {
-        s32 absDiff = ABS((s16)(*pValue - limit));
 
-        if (absDiff < 0x4000) {
-            *pValue = limit;
-            return 1;
-        }
+    if ((step * (*pValue - target)) >= 0.0f) {
+      *pValue = target;
+      return TRUE;
     }
-
-    return 0;
+  } else {
+    if (*pValue == target) {
+      return TRUE;
+    }
+  }
+  return FALSE;
 }
 
-void inter_float(f32* pValue, f32 arg1, s32 arg2) {
-    if (arg2 <= 0) {
-        *pValue = arg1;
-    } else {
-        f32 diff = arg1 - *pValue;
+/**
+ * @brief Chase an xyz_t value towards a target xyz_t value, with a specified fraction.
+ *
+ * This function updates the xyz_t value, moving it towards the target xyz_t value
+ * by a specified fraction of the distance between the current value and the target.
+ * The function returns the remaining distance after the chase.
+ *
+ * @param pValue Pointer to the xyz_t value to be updated.
+ * @param target Pointer to the target xyz_t value to chase.
+ * @param fraction Fraction of the distance to move towards the target.
+ * @return Remaining distance after the chase.
+ */
+extern f32 chase_xyz_t(xyz_t* const pValue, const xyz_t* const target,
+                       const f32 fraction) {
+  xyz_t diff;
+  f32 dist;
+  f32 stepSize;
+
+  xyz_t_sub(target, pValue, &diff);
+
+  dist = Math3DVecLength(&diff);
+  if (dist > fraction) {
+    stepSize = fraction / dist;
+    pValue->x += stepSize * diff.x;
+    pValue->y += stepSize * diff.y;
+    pValue->z += stepSize * diff.z;
+
+    return dist - fraction;
+  } else {
+    xyz_t_move(pValue, target);
+
+    return 0.0f;
+  }
+}
 
-        *pValue += diff / arg2;
+/**
+ * @brief Chase an angle value towards a limit angle with a specified step, considering wrap-around.
+ *
+ * This function updates the angle value, moving it towards the limit angle by a specified step,
+ * taking into account the angle wrap-around (0 to 65535). The function returns TRUE if the value
+ * reaches the limit, and FALSE otherwise.
+ *
+ * @param pValue Pointer to the angle value to be updated.
+ * @param limit Limit angle to chase.
+ * @param step Step value to be added to the current angle.
+ * @return TRUE if the angle reaches the limit, FALSE otherwise.
+ */
+extern int chase_angle2(s16* const pValue, const s16 limit, const s16 step) {
+  s16 prev = *pValue;
+
+  *pValue += step;
+  if (((s16)(*pValue - limit) * (s16)(prev - limit)) <= 0) {
+    s32 absDiff = ABS((s16)(*pValue - limit));
+
+    if (absDiff < 16384) {
```

Comment:
Hex is generally much clearer for this sort of thing... I think this is 0x4000? I know what that means, at least.